Fix #4409: Do not emit abstract methods in non-native JS classes.

sjrd opened the pull request:
@gzm0 Is this an appropriate way to fix this in 1.5? Throwing the exception would be moved to the `ClassDefChecker` in the future.

gzm0 commented on compiler/src/main/scala/org/scalajs/nscplugin/GenJSCode.scala:
```diff
@@ -1825,29 +1825,34 @@ abstract class GenJSCode[G <: Global with Singleton](val global: G)
-              s.tpe.params.tail.zip(sym.tpe.params).forall {
-                case (sParam, symParam) =>
-                  sParam.tpe =:= symParam.tpe
+          if (isNonNativeJSClass(currentClassSym)) {
```
Could we avoid the nesting? (maybe by checking `isAbstractMethod` twice or putting it into a lazy val)?

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -1194,9 +1194,26 @@ object Serializers {
+      } else {
+        memberDefs0
+      }
+
```
Do this inside `readMemberDefs`?

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -1194,9 +1194,26 @@ object Serializers {
+          case m @ MethodDef(_, name, _, _, _, None) =>
+            if (!hacks.use14) {
+              throw new InvalidIRException(m,
+                  "Invalid abstract method in non-native JS class")
```
Add a comment referencing #4388 that this should be there but currently can't?

gzm0 reviewed (changes requested):
>  Is this an appropriate way to fix this in 1.5? Throwing the exception would be moved to the ClassDefChecker in the future.

Absolutely. I only have minor comments.

sjrd commented:
Updated.

In addition to addressing the comments, I had to rearrange the way the hack is done not to cause the Scala 2.11 compiler to crash. The upside is that it is easier to see exactly what will disappear when we can move the check to a ClassDefChecker.

gzm0 commented on ir/shared/src/main/scala/org/scalajs/ir/Serializers.scala:
```diff
@@ -1349,8 +1349,39 @@ object Serializers {
-    def readMemberDefs(owner: ClassName, ownerKind: ClassKind): List[MemberDef] =
-      List.fill(readInt())(readMemberDef(owner, ownerKind))
+    def readMemberDefs(owner: ClassName, ownerKind: ClassKind): List[MemberDef] = {
+      val memberDefs0 = List.fill(readInt())(readMemberDef(owner, ownerKind))
```
Not sure the `0` is still necessary. But up to you.
